cli/build.go

```diff
@@ -0,0 +1,237 @@
+	if len(args) < 1 {
+		return errors.New("bug: expecting one argument")
+	}
+	o.RootDir = args[0]
+	if o.Output == "" {
+		o.Output = "client-workspace"
+	}
+	return nil
+}
+
+func (o *BuildOptions) Validate() error {
+	if _, err := os.Stat(o.RootDir); err != nil {
+		return fmt.Errorf("workspace directory %q: %v", o.RootDir, err)
+	}
+
+	if o.Push && o.Destination == "" {
+		return fmt.Errorf("destination must be set when using --push")
+
+	}
+
+	// TODO(jpower432): Validate the reference
+	return nil
+}
+
+func (o *BuildOptions) Run(ctx context.Context) error {
+	_, _ = fmt.Fprintf(o.IOStreams.Out, "Using output directory %q\n", o.Output)
+	userSpace, err := workspace.NewLocalWorkspace(o.RootDir)
+	if err != nil {
+		return err
+	}
+
+	g := graph.NewGraph()
+
+	fileIndex := make(map[string]struct{})
+	// Do the initial walk to get an index of what is in the workspace
+	err = userSpace.Walk(func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("traversing %s: %v", path, err)
+		}
+		if info == nil {
+			return fmt.Errorf("no file info")
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		fileIndex[path] = struct{}{}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Function to determine whether the
+	// data should be replaced in the template
+	tFunc := func(value interface{}) bool {
+		stringValue, ok := value.(string)
+		if !ok {
+			return false
+		}
+
+		// If the file is found in the workspace
+		// return true
+		_, found := fileIndex[stringValue]
+		return found
+	}
+
+	for path := range fileIndex {
+		_, _ = fmt.Fprintf(o.IOStreams.Out, "Adding node %s\n", path)
+		node := graph.NewNode(path)
+
+		perr := &parser.ErrInvalidFormat{}
+		buf := new(bytes.Buffer)
+		if err := userSpace.ReadObject(ctx, path, buf); err != nil {
+			return err
+		}
+		p, err := parser.ByContentType(path, buf.Bytes())
+		switch {
+		case err == nil:
+			p.AddFuncs(tFunc)
+			node.Template, node.Links, err = p.GetLinkableData(buf.Bytes())
+			if err != nil {
+				return err
+			}
+		case !errors.As(err, &perr):
+			return err
+		}
+
+		g.Nodes[node.Name] = node
+	}
+
+	for _, node := range g.Nodes {
+		for link, data := range node.Links {
+			// Currently with the parsing implementation
+			// all initial values are expected to represent
+			// the file string data present in the content.
+			// FIXME(jpower432): Making this assumption could lead
+			// to bug when trying to translate links to a graph.
+			fpath, ok := data.(string)
+			if !ok {
+				return fmt.Errorf("link %q: value should be of type string", link)
+			}
+			if err := g.AddEdge(node.Name, fpath); err != nil {
+				return err
+			}
+		}
+	}
+
+	templateBuilder := builder.NewBuilder(userSpace)
+	renderSpace, err := workspace.NewLocalWorkspace(o.Output)
+	if err != nil {
+		return err
+	}
+	if err := templateBuilder.Run(ctx, g, renderSpace); err != nil {
+		return fmt.Errorf("error building content: %v", err)
+	}
+
+	if o.Push {
```

Comment:
Do you think we want to move this to a function so we can also use it to push a pre-built artifact? I don't even know if we have such an intermediate output of the artifact but just thinking out load here.

Reply:
I played around with the idea of separating build and push into different sub-commands. I think the workflow you are suggestion here could help us support clients like `dnf` or `yum` which would need an intermediary step between build and push. I will take that as an action on a subsequent PR.